pkg/server/span_stats.go

```diff
@@ -0,0 +1,285 @@
+	conn, err := rpcConnection.Connect(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return serverpb.NewSpanStatsClient(conn), nil
+}
+
+type uniqueBucket struct {
+	sp            *roachpb.Span
+	batchRequests uint64
+	nBytes				uint64
+}
+
+func (s *spanStatsServer) GetNodeSpanStats(
+	ctx context.Context, req *serverpb.GetNodeSpanStatsRequest,
+) (*serverpb.GetNodeSpanStatsResponse, error) {
+
+	uniqueBuckets := make(map[string]uniqueBucket)
+
+	if err := s.server.node.stores.VisitStores(func(store *kvserver.Store) error {
+		store.VisitSpanStatsBuckets(func(span *roachpb.Span, batchRequests uint64, nBytes uint64) {
+			spanAsString := span.String()
+			if stat, ok := uniqueBuckets[spanAsString]; ok {
+				stat.batchRequests += batchRequests
+			} else {
+				uniqueBuckets[spanAsString] = uniqueBucket{
+					sp:            span,
+					batchRequests: batchRequests,
+					nBytes: nBytes,
+				}
+			}
+		})
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	res := serverpb.GetNodeSpanStatsResponse{}
+	res.Stats = make([]*serverpb.GetNodeSpanStatsResponse_NodeSpanStat, 0)
+
+	for _, bucket := range uniqueBuckets {
+		res.Stats = append(res.Stats, &serverpb.GetNodeSpanStatsResponse_NodeSpanStat{
+			Span:          bucket.sp,
+			BatchRequests: bucket.batchRequests,
+			NBytes: bucket.nBytes,
+		})
+	}
+
+	return &res, nil
+}
+
+// GetSpanStatistics implements the SpanStatsServer interface.
+// Get span statistics from all stores on this node.
+// this will be called by the tenant's job.
+func (s *spanStatsServer) GetSpanStatistics(
+	ctx context.Context, req *serverpb.GetSpanStatisticsRequest,
+) (*serverpb.GetSpanStatisticsResponse, error) {
+
+	responsesByNodeID := make(map[roachpb.NodeID]serverpb.GetNodeSpanStatsResponse)
+
+	dialFn := func(ctx context.Context, nodeID roachpb.NodeID) (interface{}, error) {
+		client, err := s.dialNode(ctx, nodeID)
+		return client, err
+	}
+
+	nodeFn := func(ctx context.Context, client interface{}, nodeID roachpb.NodeID) (interface{}, error) {
+		spanStats := client.(serverpb.SpanStatsClient)
+		// TODO: update bucket boundaries before returning.
+		// TODO: use a time window.
+		nodeSpanStats, err := spanStats.GetNodeSpanStats(ctx, &serverpb.GetNodeSpanStatsRequest{})
+		// TODO check for error
+		spanStats.SetSpanBoundaries(ctx, &serverpb.SetSpanBoundariesRequest{})
+		return nodeSpanStats, err
+	}
+
+	responseFn := func(nodeID roachpb.NodeID, resp interface{}) {
+		// TODO: this name is too similar to `SpanStatsRequest` and `SpanStatsResponse`
+		nodeResponse := resp.(*serverpb.GetNodeSpanStatsResponse)
+		responsesByNodeID[nodeID] = *nodeResponse
+	}
+
+	errorFn := func(nodeID roachpb.NodeID, err error) {
+		log.Errorf(ctx, "COULD NOT GET SPAN STATS FOR NODE %d: %s", nodeID, err.Error())
+	}
+
+	err := s.server.status.iterateNodes(ctx, "iterating nodes for span stats", dialFn, nodeFn, responseFn, errorFn)
+
+	if err != nil {
+		panic("could not iterate nodes")
```

Comment:
See comment above around stray panics. Also generally prefer log.Fatals, which captures a few more things because of the passed in context, and renders nicely in our logs (easily greppable too).